consensus/fork.go

```diff
@@ -4,119 +4,115 @@ import (
-func (s *State) invalidateNode(node *blockNode) {
-	for i := range node.children {
-		s.invalidateNode(node.children[i])
+// applyUntilNode will successively apply the blocks between the state's
+// currentPath and 'bn'.
+func (s *State) applyUntilNode(bn *blockNode) (err error) {
+	// Backtrack to the common parent of 'bn' and currentPath.
+	newPath := s.backtrackToBlockchain(bn)
+
+	// Apply new nodes.
+	for _, node := range newPath[1:] {
+		// If the diffs for this node have already been generated, apply diffs
+		// directly instead of generating them. This is much faster.
+		if node.diffsGenerated {
+			s.commitDiffSet(node, DiffApply)
+		} else {
+			err = s.generateAndApplyDiff(node)
+			if err != nil {
+				break
+			}
+		}
 	}
 
-	delete(s.blockMap, node.block.ID())
-	s.badBlocks[node.block.ID()] = struct{}{}
+	return
 }
 
-// forkBlockchain will take the consensus of the State from whatever node it's
-// currently on to the node presented. An error will be returned if any of the
-// blocks that get applied in the transition are found to be invalid. If an
-// error is returned, forkBlockchain will bring the consensus variables back to
-// how they were before the call was made.
+// forkBlockchain will move the consensus set onto the 'newNode' fork. An error
+// will be returned if any of the blocks applied in the transition are found to
+// be invalid. forkBlockchain is atomic; the State is only updated if the
+// function returns nil.
 func (s *State) forkBlockchain(newNode *blockNode) (err error) {
-	// Get the state hash before attempting a fork.
-	var stateHash crypto.Hash
+	// In debug mode, record the old state hash before attempting the fork.
+	// This variable is otherwise unused.
+	var oldHash crypto.Hash
 	if DEBUG {
-		stateHash = s.stateHash()
+		oldHash = s.Hash()
 	}
 
-	// Get the list of blocks tracing from the new node to the blockchain, then
-	// rewind to the common parent.
-	backtrackNodes := s.backtrackToBlockchain(newNode)
-	commonParent := backtrackNodes[len(backtrackNodes)-1]
-	rewoundNodes := s.rewindToNode(commonParent)
+	// determine common parent
+	commonParent := s.backtrackToBlockchain(newNode)[0]
 
-	// Update the consensus to include all of the block nodes that go from the
-	// common parent to `newNode`. If any of the blocks are invalid, reverse
-	// all of the changes and switch back to the original block.
-	var appliedNodes []*blockNode
-	for i := len(backtrackNodes) - 2; i >= 0; i-- {
-		bn := backtrackNodes[i]
-		appliedNodes = append(appliedNodes, bn)
+	// save currentBlockNode in case something goes wrong
+	oldHead := s.currentBlockNode()
 
-		// If the diffs for this node have already been generated, apply diffs
-		// directly instead of generating them. This is much faster.
-		if bn.diffsGenerated {
-			direction := true // the blockNode is being applied, direction is set to true.
-			s.applyDiffSet(bn, direction)
-			continue
-		}
+	// rewind to the common parent
+	s.rewindToNode(commonParent)
```

Comment:
I do think that backtrackToBlockchain should get renamed to backtrackToCurrentPath.